### ISSUE TITLE
Threads, Forking and New Relic

### ISSUE THREAD

**Issue by jasonrclark** [resque/resque#1101]
**TL;DR;** We've got ideas for avoiding deadlocks when using New Relic and Resque together, but would like feedback on the approach, and advice on what we could contribute to Resque to improve the situation.

Long story short, threads (with locks) and forking don't mix nicely. (:wavy_dash: + :fork_and_knife:  = :collision: )
# The Problem

There are long-standing reports of deadlocking Resque jobs when using New Relic (`newrelic_rpm`). Certain updates have alleviated issues for specific users, but we still hear about problems from others--especially those running lots of jobs (think millions of jobs per day)

Some great debugging by @benweint and others has led us to suspect the problems are variations on this sequence of events:

1) New Relic starts in the Resque master process and spawns a background thread to transmit data back to our services.
2) Resque starts running normally, forking to process jobs
3) New Relic harvests once a minute and during that time takes various locks (internal locks in `newrelic_rpm`, locks in Ruby's `Logger` class, or even native locks around, for example, DNS lookup)
4) While New Relic's harvest thread has a lock, Resque forks a new job
5) Only Resque's main thread lives on in the child process, so New Relic's harvest thread isn't around to release the lock from step 3).
6) If any code tries to acquire the lock from 3), the Resque job hangs :cry: 

It seems we can't safely `fork` while holding a lock in another thread or bad things happen. If `newrelic_rpm` owned all the locks in question we could deal with them in our `after_fork` handler (and we have where we've identified them). However, at least one common case (DNS lookup) isn't even locking in Ruby-land.
# A Solution

I've pushed a branch, `resque_harvest_locking` on `newrelic_rpm` with a proposed change to our instrumentation. (https://github.com/newrelic/rpm/commit/d703271ff2638c4fa2b3edbee478a3e5c945dfd9) This synchronizes between New Relic's background thread running a harvest and Resque forking a new job (just the forking, not the full job processing). We already locked around our harvest thread, so we just sync on the same lock from a patched `Resque::Worker#fork`

This has downsides for sure:
- If the network is slow contacting New Relic, new jobs can't start until the harvest finishes.
- Ugly patching to get the lock at the right spot in Resque's processing
- The existing code doesn't look workable for Resque 2.x, so there's still work to do if this approach is taken

But in our testing it does alleviate deadlocks.
# Gathering More Information

The latest version of the `newrelic_rpm` gem ships with an executable called `nrdebug`. Given a pid, `nrdebug` drops a log file with tons of information, up to and including gdb and Ruby backtraces from the stuck process. Anyone seeing hung workers with `newrelic_rpm` should definitely run `nrdebug` before killing those workers. We'd love to see the results.
# Other Ideas?

Has anyone else found these type of issues using threads and Resque together?

Are there places we could add hooks in Resque to accomplish to make this less invasive?

Even more drastically, could there be a better approach to running plugins periodically in the Resque's master so we wouldn't need the background thread and could sidestep the locking woes?

**Thanks!**

**Comment by doliveirakn** [comment]
This ticket has been open for several years. Has there been any progress or do people just accept that having new relic with resque will cause deadlocks on occasion.

**Comment by jasonrclark** [comment]
Hey @doliveirakn! Turns out there was progress in the meantime that never got updated here. The most common issue people were hitting was with DNS resolution, as mentioned previously. Turns out as of `newrelic_rpm` 3.12.1 [we started requiring `resolv-replace`](https://github.com/newrelic/rpm/commit/41d84f2e88b6b52a544067ad219aecf1cb8e3307) in these scenarios by default. Since that time, the team has pretty much stopped hearing about the issue, which is ✨ 

While it's always a possibility to deadlock when threads and forking mix, this seems to have dealt with the primary cause of resque + New Relic pain. Are you seeing problems yourself still, or just checking in?

**Comment by doliveirakn** [comment]
Thanks for getting back. 

I came across this issue about 2 years ago and I've had new relic disabled for resque since and have been periodically checking up on this issue to see progress. 

If you think it is resolved, I will look at re-enabling newrelic and I'll report back if we hit more deadlocks.

**Comment by onyxraven** [comment]
We have had NewRelic and Resque working side by side for a while now.  We started using resolv-replace in our own code after digging into some of the stacktraces (similar to what @jasonrclark mentioned).  We don't have harvest locks on, and we havent seen a recurrance of the deadlock in a long, long time.

In othe words, I think this can be closed.

**Comment by jasonrclark** [comment]
I'm leaning toward closing as well @onyxraven. @doliveirakn could you report back if you get to try it out (or if it isn't going to happen for some reason), and I'll close then?

**Comment by doliveirakn** [comment]
I have put this live as of last week (Oct 18th)
So far, I have seen no deadlocks. This issue is likely safe to close

**Comment by jasonrclark** [comment]
Thanks @doliveirakn for reviving the discussion!
